src/nvim/tui/tui.c

```diff
@@ -527,7 +529,13 @@ static void tui_flush(UI *ui)
   }
 
   unibi_goto(ui, data->row, data->col);
+
+  if (!data->busy) {
+    unibi_out(ui, unibi_cursor_normal);
+  }
+
   flush_buf(ui);
+  unibi_out(ui, unibi_cursor_invisible);
```

Comment:
> So it seems we only need to hide the cursor when the cursor was shown here.

We need to hide the cursor most times when redrawing to avoid seeing the cursor flash around the screen. That's why the cursor is put in invisible mode after flushing: it will always be the first command send to the terminal.

We don't need to always put the cursor in invisible mode, this is the case when printing a single character at the cursor position for example. The only way I see to optimize this is to detect when we make a redraw request that would require making the cursor invisible, and if such request was made we write the cursor invisible command before flushing the output buffer. This is what the patch above did(but IMO it complicates the code for little gain)